### ISSUE TITLE
nvim_win_set_option() on immediately-closed float window affects the wrong window

### ISSUE THREAD

**Issue by Yggdroot** [neovim/neovim#11525]
- `nvim --version`: NVIM v0.4.3
- `vim -u DEFAULTS` (version: ) behaves differently?
- Operating system/version: MacOS
- Terminal name/version:
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

Suppose there are two files test.vim and test2.vim in the same directory:

test.vim
```vim
let id1 = nvim_open_win(bufadd('test.vim'), 1, {'relative':'win', 'width':20, 'height':5, 'row':15, 'col':15})
let id2 = nvim_open_win(bufadd('test2.vim'), 0, {'relative':'win', 'width':20, 'height':5, 'row':5, 'col':5})
echom id1 id2

call nvim_win_set_option(id2, 'cursorline', v:true)
call nvim_win_set_option(id2, 'winhighlight', 'Normal:Error')

call nvim_win_close(id1, 1)
call nvim_win_close(id2, 1)

edit test2.vim
```
test2.vim, the content can be anything.
```vim
let id1 = nvim_open_win(bufadd('test.vim'), 1, {'relative':'win', 'width':20, 'height':5, 'row':15, 'col':15})
let id2 = nvim_open_win(bufadd('test2.vim'), 0, {'relative':'win', 'width':20, 'height':5, 'row':5, 'col':5})
echom id1 id2

call nvim_win_set_option(id2, 'cursorline', v:true)
call nvim_win_set_option(id2, 'winhighlight', 'Normal:Error')

call nvim_win_close(id1, 1)
call nvim_win_close(id2, 1)

edit test2.vim
```

steps:
1. nvim -u NORC test.vim
2. `so %`

### Actual behaviour
1. the cursorline appears
2. the color of the window is the same as `Error`

### Expected behaviour
1. no cursorline
2. the color of the window does not change

**Comment by justinmk** [comment]
Confirmed: the main window has its options changed. If the `nvim_win_close` calls are removed, the problem doesn't happen.

Seems like a order-of-operations bug: since the floating windows are closed immediately, by the time the redraw is "resolved", the options are applied to the current window instead...?

**Comment by bfredl** [comment]
For better and worse, this is how window options in vim and neovim works. The default options for a buffer is set by the current or last window that displayed that buffer. You would have observed exactly the same with a temporary split window.

**Comment by Yggdroot** [comment]
@bfredl I don't think this issue can be closed. This is not how window options in vim and neovim works.
A option that local to window is bound to (winid, buffer number) pair, the option set for (winid1, buffer 1) should not take effect for (winid2, buffer 1).

**Comment by bfredl** [comment]
> This is not how window options in vim and neovim works.
A option that local to window is bound to (winid, buffer number) pair, the option set for (winid1, buffer 1) should not take effect for (winid2, buffer 1).

Where does this come from? Documentation? Some experiments you did? Forget "intuition" or "mental model", it is likely wrong. I've already revised my mental model for window options five times (or rather four, one time I actually convinced bram to make the behaviour slightly more consistent instead) 

I just tested again with split windows and it behaves exactly the same. While the documentation doesn't give a complete operational description how local options work, there is nothing that directly contradictory with the docs here.

**Comment by bfredl** [comment]
But as far for _adding_ a new option mode to keep options private to a window (and never have them leak to another window), we could consider that. Though there could be some weird interactions with autocmds, because `FileType` expects to be able to set local window options _once_, because of this copying behevior (we could work around it using the autocmd window, and such..) So this would be a feature request and not a bug then.

**Comment by Yggdroot** [comment]
> Where does this come from? Documentation? Some experiments you did? Forget "intuition" or "mental model", it is likely wrong. I've already revised my mental model for window options five times (or rather four, one time I actually convinced bram to make the behaviour slightly more consistent instead)
> 

Some discussions in this issue.  doc is referred by the discussion. @lacygoill

> I just tested again with split windows and it behaves exactly the same. 

How did you test it? Could you describe the exact steps please?

**Comment by bfredl** [comment]
> Some discussions in this issue.

Exactly how the presence of "some discussion" with multiple different standpoints a conclusive argument for anything?

> doc is referred by the discussion.

This is beyond useless. If you have an useful reference you should quote it _here and now_. Why should I waste my time and scan a multi-page discussion for some reference that maybe or maybe not supports your case?

> How did you test it? Could you describe the exact steps please?

e file1
sp file2
set statusline
quit
e file2

**Comment by Yggdroot** [comment]
> e file1
> sp file2
> set statusline
> quit
> e file2

Are you serious? Can your test demonstrate anything?
`set statusline` can only print the statusline, even though you mean `set statusline=`, at least you should use `setlocal`.

**Comment by bfredl** [comment]
Sorry, I meant `cursorline` but either set or setl is fine. (statusline is global-local and thus has another set of different behaviors...)

**Comment by lacygoill** [comment]
> Forget "intuition" or "mental model", it is likely wrong.

I don't think it's wrong, but it's incomplete.  Anyway, it doesn't change the fact that I think you're right on this issue.

> I've already revised my mental model for window options five times

I can relate to that; I had to do the same thing multiple times.

> (or rather four, one time I actually convinced bram to make the behaviour slightly more consistent instead)

Yes, I think it was [this PR](https://github.com/vim/vim/pull/2336).  It was a welcome one, because it made the way window-local options are applied a little more consistent.

---

I don't have the time right now to write a full explanation (it would be very long to do), and I'm not sure anyone would be interested anyway.

The way window-local options are applied is explained – in part – at `:h local-options`.
But the explanation is incomplete, and there are exceptions as well as bugs.

Here is one rule which is not completely explained at `:h local-options`, and which I inferred after many tests.  It's about the initialization of a window-local option.  There are 4 cases to consider:

   1. the buffer has never been displayed anywhere

   2. the buffer has already been displayed in at least one other window; the new window results from a split

   3. the buffer has already been displayed in at least one other window; at least one of them has been closed

   4. the buffer has already been displayed in at least one other window; none of them has been closed

In case `1.`, the local value is initialized from the global one.
In case `2.`, it's simply copied from the window which was split.
In case `3.`, it's initialized with the value it had in the **last** closed window (where it was displayed).
In case `4.`, it's initialized with the value it has in the window where it was displayed for the **first** time.

I'm not sure to have correctly understood the original post – so maybe I'm wrong – but I think the behavior which is reported is an example of the case 3.

**Comment by justinmk** [comment]
> Here is one rule which is not completely explained at `:h local-options`

given all that's written here and https://github.com/vim/vim/issues/4945 it would be nice if someone sent a doc PR (with less verbosity and more clarity than whatever the existing doc has).

**Comment by chrisbra** [comment]
I agree, a improved documentation is always welcomed at the vim project.

**Comment by bfredl** [comment]
@lacygoill Right, that seems to be consistent with the implementation `find_wininfo()` and `get_winopts()`. The order in which entries are added to `buf->b_wininfo` seems somewhat arbitrary (it is created for first window upon _opening_ because it has to be non-NULL I think, otherwise a new entry is only put at the front when _leaving_ a buffer from a window), though I wonder if any small change to the defaults would be unambiguously better (perhaps favouring a window where the buffer is currently open over a closed one, but not sure it matters to anyone).
